src/compaction/mod.rs

```diff
@@ -308,6 +227,123 @@ where
         Ok(())
     }
 
+    fn next_level_scopes<'a>(
+        version: &'a Version<R, FP>,
+        min: &mut &'a <R as Record>::Key,
+        max: &mut &'a <R as Record>::Key,
+        level: usize,
+        meet_scopes_l: &[&'a Scope<<R as Record>::Key>],
+    ) -> Result<(Vec<&'a Scope<<R as Record>::Key>>, usize, usize), CompactionError<R>> {
+        let mut meet_scopes_ll = Vec::new();
+        let mut start_ll = 0;
+        let mut end_ll = 0;
+        {
+            if !version.level_slice[level + 1].is_empty() {
+                *min = &meet_scopes_l
+                    .first()
+                    .ok_or(CompactionError::EmptyLevel)?
+                    .min;
+                *max = &meet_scopes_l
+                    .iter()
+                    .last()
+                    .ok_or(CompactionError::EmptyLevel)?
+                    .max;
+
+                start_ll = Version::<R, FP>::scope_search(min, &version.level_slice[level + 1]);
+                end_ll = Version::<R, FP>::scope_search(max, &version.level_slice[level + 1]);
+
+                let next_level_len = version.level_slice[level + 1].len();
+                for scope in version.level_slice[level + 1]
+                    [start_ll..cmp::min(end_ll + 1, next_level_len - 1)]
+                    .iter()
+                {
+                    if scope.contains(min) || scope.contains(max) {
+                        meet_scopes_ll.push(scope);
+                    }
+                }
+            }
+        }
+        Ok((meet_scopes_ll, start_ll, end_ll))
+    }
+
+    fn this_level_scopes<'a>(
+        version: &'a Version<R, FP>,
+        min: &<R as Record>::Key,
+        max: &<R as Record>::Key,
+        level: usize,
+    ) -> Option<(Vec<&'a Scope<<R as Record>::Key>>, usize, usize)> {
+        let mut meet_scopes_l = Vec::new();
+        let start_l = Version::<R, FP>::scope_search(min, &version.level_slice[level]);
+        let mut end_l = start_l;
+        {
+            for scope in version.level_slice[level][start_l..].iter() {
+                if scope.contains(min) || scope.contains(max) {
+                    meet_scopes_l.push(scope);
+                    end_l += 1;
+                } else {
+                    break;
+                }
+            }
+            if meet_scopes_l.is_empty() {
+                return None;
+            }
+        }
+        Some((meet_scopes_l, start_l, end_l))
+    }
+
+    async fn build_tables<'scan>(
+        option: &DbOption,
+        version_edits: &mut Vec<VersionEdit<<R as Record>::Key>>,
+        level: usize,
+        streams: Vec<ScanStream<'scan, R, FP>>,
+    ) -> Result<(), CompactionError<R>> {
+        let mut stream = MergeStream::<R, FP>::from_vec(streams).await?;
+
+        // Kould: is the capacity parameter necessary?
+        let mut builder = R::Columns::builder(8192);
```

Comment:
0 could be better if there is not a number, or we could provider a new method to create 0 capacity record batch builder.